src/app/modules/ibc.rs

```diff
@@ -399,180 +444,344 @@ impl<S: Store> ConnectionKeeper for Ibc<S> {
     }
 
-    fn connection_end(&self, _connection_id: &ConnectionId) -> Result<ConnectionEnd, ChannelError> {
-        todo!()
+    fn connection_end(&self, connection_id: &ConnectionId) -> Result<ConnectionEnd, ChannelError> {
+        ConnectionReader::connection_end(self, connection_id)
+            .map_err(ChannelError::ics03_connection)
     }
 
     fn connection_channels(
         &self,
-        _cid: &ConnectionId,
+        cid: &ConnectionId,
     ) -> Result<Vec<(PortId, ChannelId)>, ChannelError> {
-        todo!()
+        let path = "channelEnds".to_owned().try_into().unwrap(); // safety - path must be valid since ClientId is a valid Identifier
+        let keys = self.store.get_keys(&path);
+        let channels = keys
+            .into_iter()
+            .filter_map(|path| {
+                let value = self.store.get(Height::Pending, &path)?;
+                let channel_end: ChannelEnd = IbcRawChannelEnd::decode(value.as_slice())
+                    .map(|v| v.try_into().unwrap()) // safety - data on the store is assumed to be well-formed
+                    .ok()?;
+
+                if channel_end.connection_hops.first() == Some(cid) {
+                    if let Ok(IbcPath::ChannelEnds(port_id, channel_id)) = path.try_into() {
+                        return Some((port_id, channel_id));
+                    }
+                }
+
+                None
+            })
+            .collect();
+        Ok(channels)
     }
 
-    fn client_state(&self, _client_id: &ClientId) -> Result<AnyClientState, ChannelError> {
-        todo!()
+    fn client_state(&self, client_id: &ClientId) -> Result<AnyClientState, ChannelError> {
+        ConnectionReader::client_state(self, client_id).map_err(ChannelError::ics03_connection)
     }
 
     fn client_consensus_state(
         &self,
-        _client_id: &ClientId,
-        _height: IbcHeight,
+        client_id: &ClientId,
+        height: IbcHeight,
     ) -> Result<AnyConsensusState, ChannelError> {
-        todo!()
+        ConnectionReader::client_consensus_state(self, client_id, height)
+            .map_err(ChannelError::ics03_connection)
     }
 
     fn authenticated_capability(&self, _port_id: &PortId) -> Result<Capability, ChannelError> {
-        todo!()
+        // TODO(hu55a1n1): Copy SDK impl
+        Ok(Capability::default())
     }
 
     fn get_next_sequence_send(
         &self,
-        _port_channel_id: &(PortId, ChannelId),
+        port_channel_id: &(PortId, ChannelId),
     ) -> Result<Sequence, ChannelError> {
-        todo!()
+        let path = IbcPath::SeqSends(port_channel_id.0.clone(), port_channel_id.1.clone());
+        self.store
+            .get(Height::Pending, &path.into())
+            .map(|v| serde_json::from_str(&String::from_utf8(v).unwrap()).unwrap()) // safety - data on the store is assumed to be well-formed
+            .ok_or_else(ChannelError::implementation_specific)
     }
```

Comment:
Just thinking about packet sequences in general in light of our previous conversation about Tendermint app concurrency, this for me makes me think that reading from/writing to a single channel/port will have to be synchronous, right? I mean, we could still write to multiple channel/port combinations in parallel, but a single channel/port will have to only be accessible from a single thread/coroutine?

Not something actionable right now, of course, just trying to think of the longer-term concurrency implications.

Reply:
This is an important observation and we should definitely keep this in mind. In the current design, modules have exclusive write access to their module stores. These stores are shared with the gRPC query services which only require read access. None of the message gRPC services have been implemented yet. So, although module implementations are required to be MT-safe, the module stores are never written in parallel because (AFAIK) tendermint ABCI `deliverTx` is never called in parallel.